### PR TITLE
Add RDNA gfx1151 ROCm target support

### PR DESCRIPTION
**Pull request by lhl**
I saw there was a [recent PR for RDNA3/3.5](https://github.com/tile-ai/tilelang/pull/2044) including for Strix Halo but I had some problems w/ running things and it and turns out I had to fix some things to get it working.

- Normalize ROCm arch strings, including feature suffixes such as `gfx1151:sramecc+:xnack-`.
- Prefer HIP target selection when running under ROCm PyTorch, even if CUDA tooling is also discoverable.
- Set ROCm target attrs consistently:
  - `mcpu=gfx1151`
  - `mtriple=amdgcn-amd-amdhsa-hcc`
  - `thread_warp_size=32` for gfx10/gfx11/gfx12
  - `thread_warp_size=64` for gfx9
- Pass `mcpu` through HIP compilation paths so generated kernels compile for the actual ROCm target arch.
- Add an `RDNA` carver device model and route RDNA HIP targets away from the CDNA path.
- Add RDNA-aware tensorcore schedule scoring, biased toward 8-wave blocks.
- Add HIP `DP4A` support:
  - uses `__builtin_amdgcn_sudot4` for RDNA3/RDNA3.5/RDNA4 targets including `gfx1151`
  - uses `__builtin_amdgcn_sdot4` on supported older AMD targets
  - keeps a scalar fallback for other HIP targets
- Add focused ROCm target tests for gfx1151 detection, wave size, RDNA classification, carver routing, and auto-target behavior.

Tested on:
- Hardware: Strix Halo / Radeon 8060S Graphics
- GPU arch reported by PyTorch: `gfx1151`
- OS: Linux x86_64
- Python: `3.12.11` in the `therock` conda/mamba environment
- TileLang commit under test: `12412659`
- Baseline parent commit: `4639c279`
- TVM: `0.23.dev0` from the TileLang submodule build
- PyTorch: `2.10.0+rocm7.13.0a20260417`
- `torch.version.hip`: `7.13.26154`
- HIP compiler: HIP `7.13.26154-ca4b97ef2c`, AMD clang `23.0.0git`
- Build mode: editable source build with ROCm enabled

Smoke Test:
- Kernel: TileLang int8 SIMT matmul, `M=N=K=512`, int8 x int8 -> int32
- Target: auto-detected HIP / gfx1151

Results after this patch:
- run 0: `0.058230 ms`, `4.610 TOPS`
- run 1: `0.055946 ms`, `4.798 TOPS`
- run 2: `0.055425 ms`, `4.843 TOPS`
- median: `0.055946 ms`, `4.798 TOPS`

Slow, but it works! (before compile just fails)

Tests Passed:

- `pre-commit install --install-hooks`
- `pre-commit run --all-files`
- `python3 -m pytest testing/python/target/test_tilelang_rocm_target.py -q`
  - `6 passed`
- `python -m py_compile ...`
- `ruff check ...`
- `ruff format --check ...`
- `git diff --check`
- Direct HIP syntax probe for `__builtin_amdgcn_sudot4` with `--offload-arch=gfx1151`

~~Fails (CDNA tests improperly scoped for RDNA...):~~

Fixed scope for CDNA tets:
- Existing AMD MFMA tests compile CDNA-style MFMA builtins and fail on gfx1151:
  `__builtin_amdgcn_mfma_f32_16x16x16f16 needs target feature mai-insts`
- The run later aborted inside:
  `testing/python/debug/test_tilelang_debug_print.py::test_debug_print_buffer`

I updated the tests that improperly target RDNA to be CDNA or in one case, gfx950-only. Tests pass for the architectures they're meant for.

## Summary by CodeRabbit

* **New Features**
  * RDNA architecture support for AMD GPUs (device model, generation-aware tensor-instruction lookup, and RDNA-aware routing/tuning).
  * DP4A dot-product support for HIP/AMD with a safe software fallback and template wrapper.
  * Improved HIP target detection and explicit architecture selection during HIP compilation.

* **Bug Fixes**
  * Safer handling when optional loop/layout annotations are absent.

* **Tests**
  * Expanded ROCm/CDNA tests: arch normalization, warp sizes, target selection/routing, and new CDNA/CUDA-or-CDNA test decorators.

### CODE REVIEW

File: src/tl_templates/hip/common.h [hunk 1]
```diff
@@ -111,6 +111,53 @@ TL_DEVICE unsigned __pack_bfloat162(const bfloat16_t x, const bfloat16_t y) {
...
+TL_DEVICE void DP4A(InDatatype *a, InDatatype *b, OutDatatype *c) {
+  int a_int;
+  int b_int;
+  int c_int;
+  __builtin_memcpy(&a_int, a, sizeof(a_int));
+  __builtin_memcpy(&b_int, b, sizeof(b_int));
+  __builtin_memcpy(&c_int, c, sizeof(c_int));
+  const int out = tl_dp4a(a_int, b_int, c_int);
+  __builtin_memcpy(c, &out, sizeof(out));
+}
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/tl_templates/hip/common.h:162]: fixes pushed

File: tilelang/carver/roller/policy/tensorcore.py [hunk 2]
```diff
@@ -256,8 +263,11 @@ def _assign_block_size(self, node: PrimFuncNode, td: TileDict, block_size: int):
         if tile[ax_m] < wmma_tile[ax_m] or tile[ax_n] < wmma_tile[ax_n]:
             # allow pad, otherwise, we can not get a valid tile shape
             return None
+        space_prod = int(np.prod(space))
+        if space_prod < warps:
+            return None
 
-        factors = factorize(np.prod(space) // warps)
+        factors = factorize(space_prod // warps)
```
[the remark this answers is not in the document]
Comment (lhl) [reply tilelang/carver/roller/policy/tensorcore.py:271]: corner-case, but fixed

**Comment by lhl** [comment]
Took a look at the failing tests and fixed w/ separate commits:

  - `462e4a0b` Scope MFMA tests to CDNA targets
    - Adds `requires_cdna`
    - Skips MFMA intrinsic/preshuffle tests on RDNA

  - `35d53bf8` Scope gfx950 copy async GEMM test
    - Marks the non-pipelined gfx950 copy.async GEMM test as `requires_gfx950`

  - `d77e26d1` Skip CUDA/CDNA GEMM dtype tests on RDNA
    - Adds `requires_cuda_or_cdna`
    - Skips generic GEMM `int8`/`float32` matrix-core dtype tests on RDNA, where current WMMA support is f16/bf16

  - `355ffe28` Reject non-divisible warp layouts
    - Addresses CodeRabbit feedback by rejecting tensorcore layouts where `space_prod % warps != 0` before factorization

  Validation:
  - MFMA/gfx950 scoping behaves as expected on RDNA
  - Kernel GEMM file: `3 passed, 13 skipped`
  - Runtime-looking `ndim expected 2, got ...` repros pass when `therock` has a usable host compiler path

File: tilelang/carver/roller/policy/tensorcore.py [hunk 3]
```diff
@@ -212,6 +212,13 @@ def check_tile_shape_isvalid(self, td: TileDict):
                     return False
         return super().check_tile_shape_isvalid(td)
 
+    def score_block_size(self, n):
+        base_score = super().score_block_size(n)
+        if getattr(self.arch, "platform", None) == "RDNA":
+            warps = (n + self.arch.warp_size - 1) // self.arch.warp_size
+            return (0 if warps == 8 else 1, abs(warps - 8), *base_score)
```
[the remark this answers is not in the document]
Comment (lhl) [reply tilelang/carver/roller/policy/tensorcore.py:220]: • Not a real correctness issue.

  CodeRabbit’s premise is off: a 225-thread block does not “actually use 7 warps” for occupancy/resource purposes. It launches as
  `ceil(225 / 32) == 8` warps, with the last warp partially filled.

  Also, for this TensorCore/RDNA path, invalid non-warp-multiple candidates are already rejected before layout generation:

  `reference/tilelang/tilelang/carver/roller/policy/tensorcore.py:251`
```
  if block_size % self.arch.warp_size != 0:
      return None
```
  So for any emitted valid candidate, floor and ceil are identical. For invalid candidates, they may appear in the sorted list, but assign_block_size skips failed assignments and keeps going.

**Review by zhangnju — DISMISSED** [review]
Thanks for your contribution.

File: src/op/parallel.cc
Comment (zhangnju) [inline src/op/parallel.cc:1]: it will be safer.

File: src/tl_templates/hip/common.h
Comment (zhangnju) [inline src/tl_templates/hip/common.h:1]: please double check the SDOT4 support for some old AMD GPUs ,especially __gfx1010__ ,__gfx1011__ ,__gfx1012__  and __gfx906__. Since your PR is for Strix Halo, and the previous commits of Tilelang don't claim the support for these old GPUs, do we need to mention them here?

File: tilelang/carver/arch/rdna.py
Comment (zhangnju) [inline tilelang/carver/arch/rdna.py:1]: rdna.py will be used by both rdna3 and rdna4, so please double check whether the codes can work for RDNA4 gpus, for example, get_avaliable_tensorintrin_shapes() function. some hardcoding may have the more info to explain, like self.bandwidth: list[int] = [750, 12080]

File: tilelang/carver/roller/policy/tensorcore.py
Comment (zhangnju) [inline tilelang/carver/roller/policy/tensorcore.py:1]: Fix the platform == "RDNA" string check in score_block_size() → use is_rdna_arch().

File: tilelang/carver/arch/rdna.py
Comment (zhangnju) [inline tilelang/carver/arch/rdna.py:1]: Consider a generation-aware tensor instruction lookup in get_avaliable_tensorintrin_shapes() for future RDNA4 generality

**Review by zhangnju — COMMENTED** [review]
Thanks for your contributions to AMD Tilelang. This PR adds first-class support for AMD RDNA GPUs (specifically gfx1151/Strix Halo) in TileLang. The work is well-structured across the layers: target detection, device modeling, and code generation. But because some codes of this PR will be used by RDNA4, we may also need to check whether the codes can work well for RDAN4, like some hardcoding.

**Comment by LeiWang1999** [comment]
@lhl Thanks for you contributions! would you mind resolving those comments?

**Comment by lhl** [comment]
> @lhl Thanks for you contributions! would you mind resolving those comments?

I'll take a look soon but, and I can add guards to make sure it only supports the GPUs I have (gfx1100, gfx1151) but note: I don't have or have access to RNDA4 so besides checking that it's *not* affecting it, I won't be doing any RDNA4 optimization and will leave that for those that want to do that. I'll scope specifically to RDNA3/3.5

File: src/op/parallel.cc
Comment (lhl) [reply src/op/parallel.cc:1]: I tightened the scope so this PR’s new RDNA enablement paths are gfx11-only for now.

The tested support scope remains RDNA3/RDNA3.5: gfx1100 and gfx1151. The new RDNA carver/device model and new gfx11-specific optimized code paths no longer apply to gfx12/RDNA4. 

Existing gfx12 code paths are left unchanged, but this PR does not claim RDNA4 runtime support or tuning correctness.

File: src/tl_templates/hip/common.h
Comment (lhl) [reply src/tl_templates/hip/common.h:1]: The guards I just pushed ensure that this patch is scoped only to the tested gfx11 and will not affect gfx10 or gfx9 so I don't think need to be mentioned.

File: tilelang/carver/arch/rdna.py
Comment (lhl) [reply tilelang/carver/arch/rdna.py:1]: The guard now scopes the changes only to RDNA3/RDNA3.5 so this is no longer a problem.

File: tilelang/carver/roller/policy/tensorcore.py
Comment (lhl) [reply tilelang/carver/roller/policy/tensorcore.py:1]: fixed

File: tilelang/carver/arch/rdna.py
Comment (lhl) [reply tilelang/carver/arch/rdna.py:1]: RDNA.get_avaliable_tensorintrin_shapes() now uses a generation-keyed lookup table. Only gfx11 is populated, so the PR stays scoped to RDNA3/RDNA3.5 while leaving a clear extension point for future gfx12 work.

**Comment by lhl** [comment]
@LeiWang1999 scope is guarded to gfx11 and all feedback is resolved now

File: src/tl_templates/hip/common.h [hunk 4]
```diff
@@ -111,6 +111,56 @@ TL_DEVICE unsigned __pack_bfloat162(const bfloat16_t x, const bfloat16_t y) {
...
+#endif
+
+#if defined(__gfx906__) || defined(__gfx908__) || defined(__gfx90a__) ||       \
+    defined(__gfx940__) || defined(__gfx941__) || defined(__gfx942__) ||       \
+    defined(__gfx950__) || defined(__gfx1010__) || defined(__gfx1011__) ||     \
+    defined(__gfx1012__) || defined(__gfx1030__) || defined(__gfx1031__) ||    \
+    defined(__gfx1032__) || defined(__gfx1034__) || defined(__gfx1035__) ||    \
+    defined(TL_AMDGPU_HAS_SUDOT4)
+#define TL_AMDGPU_HAS_SDOT4 1
+#endif
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/tl_templates/hip/common.h:136]: fixed

File: tilelang/carver/arch/rdna.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        self.transaction_size: list[int] = [32, 128]
+
+        # Keep the same units as the existing CUDA/CDNA heuristic. Strix Halo
+        # is a UMA part, so use a conservative global-memory score seed.
+        self.bandwidth: list[int] = [750, 12080]
+        self.available_tensor_instructions: list[TensorInstruction] | None = None
+
+    def get_avaliable_tensorintrin_shapes(self):
+        self.available_tensor_instructions = (TensorInstruction("wmma", [16, 16]),)
+        return [t.shape for t in self.available_tensor_instructions]
```
[the remark this answers is not in the document]
Comment (lhl) [reply tilelang/carver/arch/rdna.py:58]: type-alignment fixed

File: tilelang/carver/arch/__init__.py [hunk 6]
```diff
@@ -18,6 +20,8 @@ def get_arch(target: str | Target = "cuda") -> TileDevice:
     elif target.kind.name == "llvm":
         return CPU(target)
     elif target.kind.name == "hip":
+        if target_is_rdna(target) and target_get_rdna_generation(target) == 11:
+            return RDNA(target)
         return CDNA(target)
```
[the remark this answers is not in the document]
Comment (lhl) [reply tilelang/carver/arch/__init__.py:27]: non-gfx11 RDNA already silently fall to CDNA path as existing behavior. Our patch is scoped only to gfx11 however since an RDNA path is now explicitly added, I've changed to an explicit failure path to reject unsupported RDNA (instead of the previous failing once it went down the CDNA path) as the fix.

**Comment by lhl** [comment]
@LeiWang1999 @zhangnju FYI all manual and automated issues are reviewed/resolved. This patch is tightly scoped to gfx11.

**Comment by zhangnju** [comment]
HI  @lhl , This branch has conflicts that must be resolved, could you help resolve the conflicts?

**Comment by lhl** [comment]
> HI @lhl , This branch has conflicts that must be resolved, could you help resolve the conflicts?

There weren’t any conflicts when I last left it 5 days ago. I can do a pass, but only if you guys are serious about taking upstream PRs.

**Review by zhangnju — APPROVED** [review]

**Comment by zhangnju** [comment]
> > HI @lhl , This branch has conflicts that must be resolved, could you help resolve the conflicts?
> 
> There weren’t any conflicts when I last left it 5 days ago. I can do a pass, but only if you guys are serious about taking upstream PRs.

@lhl   it is my bad, last time I had approved it and just left some comments for you to modify. I thought once you completed the modifications, the PR could be merged automatically.  please continue to contribute your PRs. Thanks

**Review by LeiWang1999 — APPROVED** [review]
